File: FunctionalProgramming/SwapNodes/swap_nodes.ml

let rec swap lc rc cur h k =
	if h mod k = 0 then
	begin
		let tmp = lc.(cur) in
			lc.(cur) <- rc.(cur);
			rc.(cur) <- tmp
	end;
	if lc.(cur) != -1 then swap lc rc lc.(cur) (h+1) k;
	if rc.(cur) != -1 then swap lc rc rc.(cur) (h+1) k
;;


let rec print lc rc cur =
	if lc.(cur) != -1 then print lc rc lc.(cur);
	Printf.printf "%d " cur;
	if rc.(cur) != -1 then print lc rc rc.(cur)
;;

let () =
	let n = Scanf.scanf " %d " (fun x -> x) in
	let lc = Array.make (n+1) 0 in
	let rc = Array.make (n+1) 0 in
		for i = 1 to n do
			let l = Scanf.scanf " %d " (fun x -> x) in
			let r = Scanf.scanf " %d " (fun x -> x) in
				lc.(i) <- l;
				rc.(i) <- r
		done;
		let times = Scanf.scanf " %d " (fun x -> x) in
		let rec loop t =
			if t > 0 then
			begin
				let k = Scanf.scanf " %d " (fun x -> x) in
				swap lc rc 1 1 k;
				print lc rc 1;
				Printf.printf "\n";
				loop (t-1)
			end
		in loop times
;;
			

File: DataStructures/Trees/SwapNodes/swap_nodes.go
package main

import (
	"fmt"
)

var l, r map[int]int

func swap(cur, h, k int) {
	if cur == -1 {
		return
	}

	if h%k == 0 {
		l[cur], r[cur] = r[cur], l[cur]
	}

	swap(l[cur], h+1, k)
	swap(r[cur], h+1, k)
}

func inorder(cur int) {
	if cur == -1 {
		return
	}

	inorder(l[cur])
	fmt.Printf("%d ", cur)
	inorder(r[cur])
}

func main() {
	var n, T, a, b, k int
	l, r = make(map[int]int), make(map[int]int)

	fmt.Scanf("%d", &n)
	for i := 1; i <= n; i++ {
		fmt.Scanf("%d %d", &a, &b)
		l[i] = a
		r[i] = b
	}
	fmt.Scanf("%d", &T)
	for i := 0; i < T; i++ {
		fmt.Scanf("%d", &k)
		swap(1, 1, k)
		inorder(1)
		fmt.Println()
	}
}
